Library/OcCpuLib/OcCpuLib.c

```diff
@@ -661,46 +663,70 @@ ScanIntelProcessor (
   // SkyLake and later have an Always Running Timer
   //
   if (Cpu->Model >= CPU_MODEL_SKYLAKE) {
-    AsmCpuid (CPUID_TIME_STAMP_COUNTER, &CpuidEax, &CpuidEbx, NULL, NULL);
+    AsmCpuid (CPUID_TIME_STAMP_COUNTER, &CpuidEax, &CpuidEbx, &CpuidEcx, NULL);
+    if (CpuidEcx > 0) {
+      Cpu->ARTFrequency = CpuidEcx;
+      DEBUG ((DEBUG_INFO, "OCCPU: Core Crystal Clock Frequency %u\n", CpuidEcx));
+    } else {
+      //
+      // Fall back to identifying ART frequency based on model
+      //
+      if (Cpu->Family == 0x6 && Cpu->Model == CPU_MODEL_SKYLAKE_W) {
+        //
+        // TODO: Xeon Ws share the same CPUID but I believe they have 24 Mhz ART
+        //
+        Cpu->ARTFrequency = 25000000ULL; // 25 Mhz
+      } else if (Cpu->Family == 0x6 && Cpu->Model == CPU_MODEL_GOLDMONT){
+        Cpu->ARTFrequency = 19200000ULL; // 19.2 Mhz
+      } else {
+        Cpu->ARTFrequency = 24000000ULL; // 24 Mhz
+      }
+    }
 
     if (CpuidEax > 0 && CpuidEbx > 0) {
-      Cpu->CPUFrequency = MultU64x32 (BASE_ART_CLOCK_SOURCE, (UINT32) DivU64x32 (CpuidEbx, CpuidEax));
+      TscAdjust = AsmReadMsr64 (MSR_IA32_TSC_ADJUST);
+      DEBUG ((DEBUG_INFO, "OCCPU: TSC Adjust %llu\n", TscAdjust ));
+      ASSERT (Cpu->ARTFrequency > 0ULL);
+      Cpu->CPUFrequencyFromART = MultU64x32 (Cpu->ARTFrequency, (UINT32) DivU64x32 (CpuidEbx, CpuidEax)) + TscAdjust;
 
```

Comment:
This looks wrong to me.

Firstly because the calculation may overflow, and thus I want it to rely on `MultThenDivU64x64x32` I posted [here](https://github.com/acidanthera/bugtracker/issues/448#issuecomment-519468023). Would you please add it to `OcMiscLib`?

Secondly because `TscAdjust` is in TSC values. I reread the docs, and it is just an offset to get a TSC value from ART. CPU frequency will not be affected with it, as these are two different calculations:

```
CPU frequency = ART frequency * Ebx / Eax
TSC timestamp = ART timestamp * Ebx / Eax + TSC offset
```

ART timestamps are provided by some onboard hardware. I believe the formula in `CPUFrequencyFromART` description needs to be changed for this reason as well.

Reply:
Good catch! I added `MultThenDivU64x64x32` to `OcMiscLib` in a new file called `Math.c`. I adjusted the ART-based CPU frequency calculation and adjusted the documentation to match.